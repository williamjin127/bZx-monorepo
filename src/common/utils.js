import { BigNumber } from "bignumber.js";

// const SCALE_FACTOR = BigNumber(1e18);

export const fromBigNumber = (bigNum, scale = 1) =>
  bigNum.div(scale).toNumber();

export const toBigNumber = (num, scale = 1) => {
  const bigNum = BigNumber(num);
  return bigNum.times(scale);
};
