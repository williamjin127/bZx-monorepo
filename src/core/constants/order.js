export const MAKER_ROLE = {
  LENDER: "0",
  TRADER: "1"
};
