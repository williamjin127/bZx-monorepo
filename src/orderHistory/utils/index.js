const ORDER_FIELD_COUNT = 14;
const SOMETHING = 64;

const remove0xPrefix = data => data.substr(2);
const getOrderObjArray = data =>
  data.match(new RegExp(`.{1,${ORDER_FIELD_COUNT * SOMETHING}}`, "g"));
const getOrder = data => data.match(new RegExp(`.{1,${SOMETHING}}`, "g"));

export const cleanData = raw => {
  const no0x = remove0xPrefix(raw);
  const objCount = no0x.length / SOMETHING / ORDER_FIELD_COUNT;
  if (objCount % 1 !== 0) throw new Error("Must be whole number of objects");

  const orderObjArray = getOrderObjArray(no0x);

  const hexRadix = 16;
  const orders = orderObjArray.map(orderObj => {
    const params = getOrder(orderObj);
    return {
      maker: `0x${params[0].substr(24)}`,
      loanTokenAddress: `0x${params[1].substr(24)}`,
      interestTokenAddress: `0x${params[2].substr(24)}`,
      collateralTokenAddress: `0x${params[3].substr(24)}`,
      feeRecipientAddress: `0x${params[4].substr(24)}`,
      oracleAddress: `0x${params[5].substr(24)}`,
      loanTokenAmount: parseInt(`0x${params[6]}`, hexRadix),
      interestAmount: parseInt(`0x${params[7]}`, hexRadix),
      initialMarginAmount: parseInt(`0x${params[8]}`, hexRadix),
      maintenanceMarginAmount: parseInt(`0x${params[9]}`, hexRadix),
      lenderRelayFee: parseInt(`0x${params[10]}`, hexRadix),
      traderRelayFee: parseInt(`0x${params[11]}`, hexRadix),
      expirationUnixTimestampSec: parseInt(`0x${params[12]}`, hexRadix),
      loanOrderHash: `0x${params[13]}`
    };
  });
  return orders;
};
