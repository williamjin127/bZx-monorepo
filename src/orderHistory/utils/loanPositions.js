import { pipe, map } from "ramda";
import * as Utils from "./index";

const NUM_LOAN_POS_FIELDS = 9;
const HEX_RADIX = 16;

const getLoanPosition = params => ({
  lender: `0x${params[0].substr(24)}`,
  trader: `0x${params[1].substr(24)}`,
  collateralTokenAddressFilled: `0x${params[2].substr(24)}`,
  positionTokenAddressFilled: `0x${params[3].substr(24)}`,
  loanTokenAmountFilled: parseInt(`0x${params[4]}`, HEX_RADIX),
  collateralTokenAmountFilled: parseInt(`0x${params[5]}`, HEX_RADIX),
  positionTokenAmountFilled: parseInt(`0x${params[6]}`, HEX_RADIX),
  loanStartUnixTimestampSec: parseInt(`0x${params[7]}`, HEX_RADIX),
  active: parseInt(`0x${params[8]}`, HEX_RADIX)
});

const checkProperObjCount = Utils.makeCheckProperObjCount(NUM_LOAN_POS_FIELDS);
const getOrderObjArray = Utils.makeGetOrderObjArray(NUM_LOAN_POS_FIELDS);

export const cleanData = raw =>
  pipe(
    Utils.remove0xPrefix,
    checkProperObjCount,
    getOrderObjArray,
    map(pipe(Utils.getOrderParams, getLoanPosition))
  )(raw);
