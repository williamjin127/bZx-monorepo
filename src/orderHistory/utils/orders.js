import { pipe, map } from "ramda";
import * as Utils from "./index";

const NUM_ORDER_FIELDS = 14;
const HEX_RADIX = 16;

const getOrder = params => ({
  maker: `0x${params[0].substr(24)}`,
  loanTokenAddress: `0x${params[1].substr(24)}`,
  interestTokenAddress: `0x${params[2].substr(24)}`,
  collateralTokenAddress: `0x${params[3].substr(24)}`,
  feeRecipientAddress: `0x${params[4].substr(24)}`,
  oracleAddress: `0x${params[5].substr(24)}`,
  loanTokenAmount: parseInt(`0x${params[6]}`, HEX_RADIX),
  interestAmount: parseInt(`0x${params[7]}`, HEX_RADIX),
  initialMarginAmount: parseInt(`0x${params[8]}`, HEX_RADIX),
  maintenanceMarginAmount: parseInt(`0x${params[9]}`, HEX_RADIX),
  lenderRelayFee: parseInt(`0x${params[10]}`, HEX_RADIX),
  traderRelayFee: parseInt(`0x${params[11]}`, HEX_RADIX),
  expirationUnixTimestampSec: parseInt(`0x${params[12]}`, HEX_RADIX),
  loanOrderHash: `0x${params[13]}`
});

const checkProperObjCount = Utils.makeCheckProperObjCount(NUM_ORDER_FIELDS);
const getOrderObjArray = Utils.makeGetOrderObjArray(NUM_ORDER_FIELDS);

export const cleanData = raw =>
  pipe(
    Utils.remove0xPrefix,
    checkProperObjCount,
    getOrderObjArray,
    map(pipe(Utils.getOrderParams, getOrder))
  )(raw);
