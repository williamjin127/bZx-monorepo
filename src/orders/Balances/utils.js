/* globals localStorage */

export const permaTokens = [`WETH_SM_ADDRESS_HERE`, `ZRX_SM_ADDRESS_HERE`];

export const getTrackedTokens = () => {
  const storedTokens = localStorage.getItem(`trackedTokens`);
  return storedTokens
    ? [...permaTokens, ...JSON.parse(storedTokens)]
    : permaTokens;
};

export const addTrackedToken = tokenAddress => {
  const currTokens = getTrackedTokens();
  const newTrackedTokens = currTokens.push(tokenAddress);
  return localStorage.setItem(`trackedTokens`, newTrackedTokens);
};

export const removeTrackedToken = tokenAddress => {
  const currTokens = getTrackedTokens();
  const newTrackedTokens = currTokens.filter(token => token !== tokenAddress);
  return localStorage.setItem(`trackedTokens`, newTrackedTokens);
};
