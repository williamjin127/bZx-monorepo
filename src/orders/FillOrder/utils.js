import styled from "styled-components";
import B0xJS from "b0x.js";  // eslint-disable-line
import { getTrackedTokens } from "../../common/trackedTokens";
import { toBigNumber } from "../../common/utils";

const TxHashLink = styled.a.attrs({
  target: `_blank`,
  rel: `noopener noreferrer`
})`
  font-family: monospace;
  display: block;
  text-overflow: ellipsis;
  overflow: auto;
}
`;

export const getOrderHash = order => B0xJS.getLoanOrderHashHex(order);

const checkAllowance = async (b0x, accounts, tokenAddress) => {
  const allowance = await b0x.getAllowance({
    tokenAddress,
    ownerAddress: accounts[0].toLowerCase()
  });
  return allowance.toNumber() !== 0;
};

const checkCoinsApproved = async (
  b0x,
  accounts,
  order,
  collateralTokenAddress
) => {
  const makerRole = order.makerRole === `0` ? `lender` : `trader`;
  if (makerRole === `lender`) {
    // check that user has approved collateralToken and interestToken
    const { interestTokenAddress } = order;
    const a = await checkAllowance(b0x, accounts, interestTokenAddress);
    const b = await checkAllowance(b0x, accounts, collateralTokenAddress);
    return a && b;
  }
  // check that user has approved loanToken
  const { loanTokenAddress } = order;
  const a = await checkAllowance(b0x, accounts, loanTokenAddress);
  return a;
};

export const validateFillOrder = async (
  order,
  fillOrderAmount,
  collateralTokenAddress,
  tokens,
  b0x,
  accounts
) => {
  const makerRole = order.makerRole === `0` ? `lender` : `trader`;
  const trackedTokens = getTrackedTokens(tokens);
  if (makerRole === `lender`) {
    // make sure the collateral token chosen is allowed
    const collateralToken = tokens.filter(
      t => t.address === collateralTokenAddress
    )[0];
    const notAllowed = {
      1: [],
      3: [`ZRX`, `B0X`],
      4: [],
      42: []
    };
    const collateralTokenNotAllowed = notAllowed[b0x.networkId].includes(
      collateralToken && collateralToken.symbol
    );
    if (collateralTokenNotAllowed) {
      alert(
        `The selected tokens are not yet supported for lending or collateral.`
      );
      return false;
    }

    // check that user has tracked collateralToken and interestToken
    const { interestTokenAddress } = order;
    if (
      !trackedTokens.includes(interestTokenAddress) ||
      !trackedTokens.includes(collateralTokenAddress)
    ) {
      alert(
        `Your interest token or collateral token is not tracked, please add it in the balances tab.`
      );
      return false;
    }
  } else {
    // check that user has tracked loanToken
    const { loanTokenAddress } = order;
    if (!trackedTokens.includes(loanTokenAddress)) {
      alert(
        `Your loan token is not tracked, please add it in the balances tab.`
      );
      return false;
    }
  }

  const coinsApproved = await checkCoinsApproved(
    b0x,
    accounts,
    order,
    collateralTokenAddress
  );
  if (!coinsApproved) {
    alert(
      `Some of your coins are not approved, please check the balances tab.`
    );
    return false;
  }
  console.log(`validateFillOrder`);
  console.log(order, fillOrderAmount, collateralTokenAddress);
  return true;
};
export const submitFillOrder = async (
  order,
  fillOrderAmount,
  collateralTokenAddress,
  b0x,
  accounts
) => {
  const txOpts = { from: accounts[0].toLowerCase() };
  const makerIsLender = order.makerRole === `0`;

  console.log(`order`, order);
  console.log(`txOpts`, txOpts);
  console.log(`makerIsLender`, makerIsLender);

  let receipt;
  if (makerIsLender) {
    // receipt = await b0x.takeLoanOrderAsTrader(
    //   order,
    //   collateralTokenAddress,
    //   toBigNumber(fillOrderAmount, 1e18),
    //   txOpts
    // );

    b0x
      .takeLoanOrderAsTrader(
        order,
        collateralTokenAddress,
        toBigNumber(fillOrderAmount, 1e18),
        txOpts
      )
      .once(`transactionHash`, hash => {
        alert(`Transaction submitted, transaction hash:`, {
          component: () => (
            <TxHashLink href={`${b0x.etherscanURL}tx/${hash}`}>
              {hash}
            </TxHashLink>
          )
        });
      })
      .on(`error`, error => {
        console.error(error.message);
      });
  } else {
    // receipt = await b0x.takeLoanOrderAsLender(order, txOpts);
    b0x
      .takeLoanOrderAsLender(order, txOpts)
      .once(`transactionHash`, hash => {
        alert(`Transaction submitted, transaction hash:`, {
          component: () => (
            <TxHashLink href={`${b0x.etherscanURL}tx/${hash}`}>
              {hash}
            </TxHashLink>
          )
        });
      })
      .on(`error`, error => {
        console.error(error.message);
      });
  }

  console.log(receipt);
  return true;
};
