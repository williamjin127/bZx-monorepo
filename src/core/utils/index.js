import { BigNumber } from "@0xproject/utils";
import { assert } from "@0xproject/assert";
import { constants } from "0x.js/lib/src/utils/constants";
import BN from "bn.js";
import Web3Utils from "web3-utils";
import * as Errors from "../constants/errors";
import { SchemaValidator } from "../../schemas/b0x_json_schemas";
import { ropsten as Contracts } from "../../contracts";
import * as Addresses from "../../addresses";

export const noop = () => {};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const randomNumber = BigNumber.random(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT
  );
  const factor = new BigNumber(10).pow(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT - 1
  );
  const salt = randomNumber.times(factor).round();
  return salt;
};

export const getLoanOrderHashHex = order => {
  const orderAddrs = [
    order.makerAddress,
    order.loanTokenAddress,
    order.interestTokenAddress,
    order.collateralTokenAddress,
    order.feeRecipientAddress,
    order.oracleAddress
  ];
  const orderUints = [
    bigNumberToBN(order.loanTokenAmount),
    bigNumberToBN(order.interestAmount),
    bigNumberToBN(order.initialMarginAmount),
    bigNumberToBN(order.maintenanceMarginAmount),
    bigNumberToBN(order.lenderRelayFee),
    bigNumberToBN(order.traderRelayFee),
    bigNumberToBN(order.expirationUnixTimestampSec),
    bigNumberToBN(order.makerRole),
    bigNumberToBN(order.salt)
  ];
  const orderHashHex = Web3Utils.soliditySha3(
    { t: "address", v: order.b0xAddress },
    { t: "address[6]", v: orderAddrs },
    { t: "uint256[9]", v: orderUints }
  );
  return orderHashHex;
};

export const doesContractExistAtAddress = async (web3, address) => {
  const code = await web3.eth.getCode(address);
  // Regex matches 0x0, 0x00, 0x in order to accommodate poorly implemented clients
  const codeIsEmpty = /^0x0{0,40}$/i.test(code);
  return !codeIsEmpty;
};

export const getContractInstance = async (web3, abi, address) => {
  assert.isETHAddressHex("address", address);
  const contractExists = await doesContractExistAtAddress(web3, address);
  if (!contractExists)
    throw new Error(`${Errors.ContractDoesNotExist} ${address}`);

  const contract = new web3.eth.Contract(abi, address);
  return contract;
};

export const doesConformToSchema = (variableName, value, schema) => {
  const schemaValidator = new SchemaValidator();
  const validationResult = schemaValidator.validate(value, schema);
  const hasValidationErrors = validationResult.errors.length > 0;
  const msg = `Expected ${variableName} to conform to schema ${
    schema.id
  }\nEncountered: ${JSON.stringify(
    value,
    null,
    "\t"
  )}\nValidation errors: ${validationResult.errors.join(", ")}`;
  assert.assert(!hasValidationErrors, msg);
};

export const isValidSignature = async (
  web3,
  { account, orderHash, signature }
) => {
  const b0xContract = await getContractInstance(
    web3,
    Contracts.B0x.abi,
    Addresses.getAddresses(web3.currentProvider).B0x
  );
  return b0xContract.methods
    .isValidSignature(account, orderHash, signature)
    .call();
};
