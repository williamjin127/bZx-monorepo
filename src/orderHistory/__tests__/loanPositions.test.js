import { constants as constantsZX } from "0x.js/lib/src/utils/constants";
import b0xJS from "../../core/__tests__/setup";
import Accounts from "../../core/__tests__/accounts";
import * as FillTestUtils from "../../fill/__tests__/utils";
import makeOrder from "../../core/__tests__/order";
import * as orderConstants from "../../core/constants/order";
import B0xJS from "../../core";

const { web3 } = b0xJS;

describe("loan positions", () => {
  const owner = Accounts[0].address;
  const lenders = [Accounts[5].address, Accounts[7].address];
  const traders = [Accounts[6].address, Accounts[8].address];

  beforeAll(async () => {
    const {
      loanTokens,
      collateralTokens,
      interestTokens,
      b0xToken
    } = await FillTestUtils.initAllContractInstances();
    const ownerTxOpts = { from: owner };
    const transferAmt = web3.utils.toWei("1000000", "ether");

    await FillTestUtils.setupB0xToken({
      b0xToken,
      lenders,
      traders,
      transferAmt,
      ownerTxOpts
    });
    await FillTestUtils.setupLoanTokens({
      loanTokens,
      lenders,
      transferAmt,
      ownerTxOpts
    });
    await FillTestUtils.setupCollateralTokens({
      collateralTokens,
      traders,
      transferAmt,
      ownerTxOpts
    });
    await FillTestUtils.setupInterestTokens({
      interestTokens,
      traders,
      transferAmt,
      ownerTxOpts
    });

    const makerAddress = lenders[0];
    const takerAddress = traders[0];
    const txOpts = {
      from: takerAddress,
      gas: 1000000,
      gasPrice: web3.utils.toWei("30", "gwei").toString()
    };
    const expirationUnixTimestampSec = "1719061340";

    const order = makeOrder({
      makerAddress,
      loanTokenAddress: loanTokens[0].options.address.toLowerCase(),
      interestTokenAddress: interestTokens[0].options.address.toLowerCase(),
      collateralTokenAddress: constantsZX.NULL_ADDRESS,
      feeRecipientAddress: constantsZX.NULL_ADDRESS,
      loanTokenAmount: web3.utils.toWei("251").toString(),
      interestAmount: web3.utils.toWei("2").toString(),
      initialMarginAmount: "50",
      maintenanceMarginAmount: "25",
      lenderRelayFee: web3.utils.toWei("0.001").toString(),
      traderRelayFee: web3.utils.toWei("0.0015").toString(),
      expirationUnixTimestampSec,
      makerRole: orderConstants.MAKER_ROLE.LENDER,
      salt: B0xJS.generatePseudoRandomSalt().toString()
    });

    const orderHashHex = B0xJS.getLoanOrderHashHex(order);
    const signature = await b0xJS.signOrderHashAsync(
      orderHashHex,
      makerAddress
    );

    const loanTokenAmountFilled = web3.utils.toWei("12.3");
    await b0xJS.takeLoanOrderAsTrader(
      { ...order, signature },
      collateralTokens[0].options.address.toLowerCase(),
      loanTokenAmountFilled,
      txOpts
    );
  });

  describe("getLoanPositions", async () => {
    /* NOTE: If you want to re-run this test, you must restart
     the local testnet as the orders made by this address will
      accumulate causing this test to fail */
    test("should return loan positions", async () => {
      const loanPositions = await b0xJS.getLoanPositions({
        loanPartyAddress: traders[0],
        start: 0,
        count: 10
      });

      /*
      One thing to keep in mind with tests against takeLoanOrderAsLender or takeLoanOrderAsTrader..
      to calcuate the amount of collateral token amount required and transfered, b0x does a call to the oracle to get the current exchange rate 
      (between collateralToken and loanToken), then based on that and the initialMarginAmount, 
      it calculates and transfers enough collateral token from the trader to satisfy margin requirements. 
      Since the testnet isn't connected to Kyber to get true token rates, the oracle just randomly generates a bogus rate. 
      */
      const loanPositionsNoRandomFields = loanPositions.map(
        ({ loanStartUnixTimestampSec, collateralTokenAmountFilled, ...rest }) =>
          rest
      );

      expect(loanPositionsNoRandomFields).toContainEqual({
        active: 1,
        collateralTokenAddressFilled:
          "0xf96b018e8de3a229dbaced8439df9e3034e263c1",
        lender: "0xa8dda8d7f5310e4a9e24f8eba77e091ac264f872",
        loanTokenAmountFilled: 12300000000000000000,
        positionTokenAddressFilled:
          "0x8a063452f7df2614db1bca3a85ef35da40cf0835",
        positionTokenAmountFilled: 12300000000000000000,
        trader: "0x06cef8e666768cc40cc78cf93d9611019ddcb628"
      });
    });
  });
});
