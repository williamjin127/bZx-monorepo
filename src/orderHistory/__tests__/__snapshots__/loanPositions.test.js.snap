// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loan positions getLoanPositions should return loan positions 1`] = `
Array [
  Object {
    "active": 1,
    "collateralTokenAddressFilled": "0x8a063452f7df2614db1bca3a85ef35da40cf0835",
    "collateralTokenAmountFilled": 135300000000000000000,
    "lender": "0xa8dda8d7f5310e4a9e24f8eba77e091ac264f872",
    "loanTokenAmountFilled": 12300000000000000000,
    "positionTokenAddressFilled": "0x0d8b0dd11f5d34ed41d556def5f841900d5b1c6b",
    "positionTokenAmountFilled": 12300000000000000000,
    "trader": "0x06cef8e666768cc40cc78cf93d9611019ddcb628",
  },
]
`;
